### DISCUSSION TITLE
Desktop Environment icons are not displayed

### DISCUSSION THREAD

**Discussion by ReedClanton** [gmodena/nix-flatpak#187] [Q&A] [answered]
# Description

When flatpaks are installed via the Home Manager module, Desktop Environment icons don't appear. The applications are installed under `~/.local/share/flatpak/app`, and can be launched from the terminal.

# Launch Output

Output when flatpaks are launched via the terminal:

```
Note that the directories 

'/var/lib/flatpak/exports/share'
'/home/reedclanton/.local/share/flatpak/exports/share'

are not in the search path set by the XDG_DATA_DIRS environment variable, so
applications installed by Flatpak may not appear on your desktop until the
session is restarted.
```

# Temporary Fix

Add the bellow to your Home Manager configuration:

```
xdg.systemDirs.data = [
    "/var/lib/flatpak/exports/share"
    "/home/<userName>/.local/share/flatpak/exports/share"
];
```

# Reproduce

1. Fresh NixOS install w/ Gnome done with the provided NixOS GUI installer.
2. *Only* install the Home Manger `nix-flatpak` module.

# My `nix-flatpak` Nix File

```
{ nix-flatpak, pkgs, user, ... }: {
        home.packages = with pkgs; [ flatpak ];

        services.flatpak = {
                enable = true;
                packages = [ { appId = "com.github.tchx84.Flatseal"; origin = "flathub"; } ];
                # TODO: Troubleshoot why adding a remote prevents flatpak install and add issue to nix-flatpak github.
#               remotes = [{
#                       name = "flathub-beta";
#                       location = "https://flathub.org/beta-repo/flathub-beta.flatpakrepo";
#               }];
                update = {
                        auto.enable = true;
                        onActivation = true;
                };
        };

#       xdg.systemDirs.data = [
#               "/var/lib/flatpak/exports/share"
#               "${user.home}/.local/share/flatpak/exports/share"
#       ];
}

```

**Comment by farberbrodsky** [comment, marked as answer]
interjecting myself into this issue. I do not know if there is a better technical solution but I think this is important to at least write documentation for in the installation README.md.

Under "Notes on HomeManager", another paragraph should be added:

"
In order to show desktop icons for installed apps in a home-manager installation, add the following to your home.nix:
```nix
xdg.systemDirs.data = [ "$HOME/.local/share/flatpak/exports/share" ];
```
This will add flatpaks' desktop files and icons to your session's XDG_DATA_DIRS. 
"

**Comment by Kladki** [comment]
The above workaround worked for a while, but then suddenly the apps disappeared again. I am new to NixOS, so I am not sure what could've caused it.

**Comment by Kladki** [comment]
I figured out the problem. I missed this in the README, but [you need to have `services.flatpak.enable = true` in your NixOS configuration](https://github.com/gmodena/nix-flatpak?tab=readme-ov-file#getting-started) to use the home manager module too.

**Comment by gmodena** [comment]
> interjecting myself into this issue. I do not know if there is a better technical solution but I think this is important to at least write documentation for in the installation README.md.
> 
> Under "Notes on HomeManager", another paragraph should be added:
> 
> " In order to show desktop icons for installed apps in a home-manager installation, add the following to your home.nix:
> 
> xdg.systemDirs.data = [ "$HOME/.local/share/flatpak/exports/share" ];
> 
> This will add flatpaks' desktop files and icons to your session's XDG_DATA_DIRS. "

Thanks for looking into this. I'll move this Issue to a Q&A conversation, mark your comment as the answer, and point to it in the README.

I never had to fiddle with [XDG](https://specifications.freedesktop.org/basedir/latest/) to get desktop icons to work (be it flatpaks or normal apps). Just for my own education, are you declaring `xdg.*` (`XDG_DATA_HOME``) anywhere else in your NixOS configuration?

**Reply by RealHypnoticOcelot** [reply]
I'm having this issue too, and declaring `xdg.systemDirs.data` doesn't seem to do anything. I've enabled xdg.portal, but that's all I've set in terms in XDG.
